java-frontend/src/test/java/org/sonar/java/resolve/Java10SemanticTest.java

```diff
@@ -0,0 +1,204 @@
+    assertThat(type).is("java.util.List");
+
+    assertThat(((JavaType) type).isParameterized()).isTrue();
+    ParametrizedTypeJavaType ptjt = (ParametrizedTypeJavaType) type;
+    JavaType substitution = ptjt.substitution(ptjt.typeParameters().get(0));
+    assertThat(substitution).is("java.lang.String");
+  }
+
+  /**
+   * JLS10 §14.4.1-1.c
+   */
+  @Test
+  public void var_parametrized_type_with_wildcard() {
+    VariableTree var = getLocalVariable("var c = \"x\".getClass();");
+    Type type = var.symbol().type();
+    assertThat(type).isEqualTo(var.type().symbolType());
+    assertThat(type).is("java.lang.Class");
+
+    assertThat(((JavaType) type).isParameterized()).isTrue();
+    ParametrizedTypeJavaType ptjt = (ParametrizedTypeJavaType) type;
+    JavaType substitution = ptjt.substitution(ptjt.typeParameters().get(0));
+    assertThat(substitution.isTagged(JavaType.WILDCARD)).isTrue();
+    WildCardType wildCardType = (WildCardType) substitution;
+    assertThat(wildCardType.boundType).isEqualTo(WildCardType.BoundType.EXTENDS);
+    assertThat(wildCardType.bound).is("java.lang.String");
+  }
+
+  /**
+   * JLS10 §14.4.1-1.d
+   */
+  @Test
+  public void var_type_anonymous_class() {
+    VariableTree var = getLocalVariable(
+      "var d = new A() { int myField; void myMethod() { } };\n"
+        + "d.myField = 42;\n"
+        + "d.myMethod();");
+    Type type = var.symbol().type();
+    assertThat(type).isEqualTo(var.type().symbolType());
+    assertThat(type).isNot("java.lang.Object");
+    assertThat(type).isSubtypeOf("java.lang.Object");
+
+    JavaSymbol.TypeJavaSymbol symbol = (JavaSymbol.TypeJavaSymbol) type.symbol();
+    assertThat(symbol.memberSymbols()).hasSize(4);
+
+    JavaSymbol thisIdentifier = symbol.members().lookup("this").get(0);
+    assertThat(thisIdentifier.declaration()).isNull();
+    JavaSymbol superIdentifier = symbol.members().lookup("super").get(0);
+    assertThat(superIdentifier.declaration()).isNull();
+
+    JavaSymbol myField = symbol.members().lookup("myField").get(0);
+    assertThat(myField.isVariableSymbol()).isTrue();
+    assertThat(myField.declaration()).isNotNull();
+    assertThat(myField.usages()).hasSize(1);
+
+    JavaSymbol myMethod = symbol.members().lookup("myMethod").get(0);
+    assertThat(myMethod.isMethodSymbol()).isTrue();
+    assertThat(myMethod.declaration()).isNotNull();
+    assertThat(myMethod.usages()).hasSize(1);
+  }
+
+  /**
+   * JLS10 §14.4.1-1.e
+   */
+  @Test
+  public void var_cast_intersection_type() {
+    VariableTree var = getLocalVariable("var e = (CharSequence & Comparable<String>) \"x\";");
+    Type type = var.symbol().type();
+    assertThat(type).isEqualTo(var.type().symbolType());
+
+    // FIXME SONARJAVA-2724: type of cast expression with bounds is currently resolved only to its first type,
+    // instead of the intersection type (JLS10 - §4.9).
+
+    // this assertion should be false
+    assertThat(type).is("java.lang.CharSequence");
```

Comment:
Given that you are not asserting something is true or false, the comment is a bit dodgy.